File: backend/tests/data_access.ts
import * as anchor from '@project-serum/anchor';
import { Program } from '@project-serum/anchor';
import { MMPDataAccess } from '../target/types/mmp_data_access';
import { TOKEN_PROGRAM_ID, createMint, createAccount, mintTo } from '@solana/spl-token';
import { expect } from 'chai';

describe('mmp_data_access', () => {
    const provider = anchor.AnchorProvider.env();
    anchor.setProvider(provider);

    const program = anchor.workspace.MMPDataAccess as Program<MMPDataAccess>;
    let mint: anchor.web3.PublicKey;
    let tokenAccount: anchor.web3.PublicKey;
    let accessLevel: anchor.web3.PublicKey;

    before(async () => {
        // Create mint and token account for testing
        mint = await createMint(
            provider.connection,
            provider.wallet.payer,
            provider.wallet.publicKey,
            null,
            9
        );

        tokenAccount = await createAccount(
            provider.connection,
            provider.wallet.payer,
            mint,
            provider.wallet.publicKey
        );

        // Mint some tokens for testing
        await mintTo(
            provider.connection,
            provider.wallet.payer,
            mint,
            tokenAccount,
            provider.wallet.publicKey,
            1000000000 // 1 billion tokens
        );
    });

    it('Creates an access level', async () => {
        const level = 1;
        const requiredTokens = new anchor.BN(1000000); // 1 million tokens

        accessLevel = anchor.web3.Keypair.generate().publicKey;

        await program.methods
            .createAccessLevel(level, requiredTokens)
            .accounts({
                accessLevel: accessLevel,
                admin: provider.wallet.publicKey,
                systemProgram: anchor.web3.SystemProgram.programId,
            })
            .rpc();

        const accessLevelAccount = await program.account.accessLevel.fetch(accessLevel);
        expect(accessLevelAccount.level).to.equal(level);
        expect(accessLevelAccount.requiredTokens.toNumber()).to.equal(requiredTokens.toNumber());
    });

    it('Verifies access with sufficient tokens', async () => {
        const requiredLevel = 1;

        await program.methods
            .verifyAccess(requiredLevel)
            .accounts({
                accessLevel: accessLevel,
                userTokenAccount: tokenAccount,
                user: provider.wallet.publicKey,
            })
            .rpc();

        // If we get here, the verification passed
        expect(true).to.be.true;
    });

    it('Denies access with insufficient tokens', async () => {
        const requiredLevel = 2; // Higher level than what we have

        try {
            await program.methods
                .verifyAccess(requiredLevel)
                .accounts({
                    accessLevel: accessLevel,
                    userTokenAccount: tokenAccount,
                    user: provider.wallet.publicKey,
                })
                .rpc();
            throw new Error('Should have failed');
        } catch (error) {
            expect(error).to.be.instanceOf(Error);
        }
    });
}); 

File: backend/programs/access/data_access.rs
use anchor_lang::prelude::*;
use anchor_spl::token::TokenAccount;

declare_id!("MMPA111111111111111111111111111111111111111");

#[program]
pub mod mmp_data_access {
    use super::*;

    pub fn create_access_level(
        ctx: Context<CreateAccessLevel>,
        level: u8,
        required_tokens: u64,
    ) -> Result<()> {
        let access_level = &mut ctx.accounts.access_level;
        access_level.level = level;
        access_level.required_tokens = required_tokens;
        Ok(())
    }

    pub fn verify_access(ctx: Context<VerifyAccess>, required_level: u8) -> Result<()> {
        let user_tokens = ctx.accounts.user_token_account.amount;
        let access_level = &ctx.accounts.access_level;
        
        require!(
            access_level.level <= required_level &&
            user_tokens >= access_level.required_tokens,
            ErrorCode::InsufficientAccess
        );
        Ok(())
    }
}

#[derive(Accounts)]
pub struct CreateAccessLevel<'info> {
    #[account(
        init,
        payer = admin,
        space = 100
    )]
    pub access_level: Account<'info, AccessLevel>,
    #[account(mut)]
    pub admin: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct VerifyAccess<'info> {
    pub access_level: Account<'info, AccessLevel>,
    pub user_token_account: Account<'info, TokenAccount>,
    pub user: Signer<'info>,
}

#[account]
pub struct AccessLevel {
    pub level: u8,
    pub required_tokens: u64,
}

#[error_code]
pub enum ErrorCode {
    #[msg("Insufficient access level or token balance")]
    InsufficientAccess,
} 